### PR TITLE
Updates to Otter Grade CSV

### PR DESCRIPTION
**Pull request by sean-morris**

### CODE REVIEW

File: otter/grade/containers.py [hunk 1]
```diff
@@ -218,12 +223,18 @@ def kill_container():
             scores = dill.load(f)
 
         scores_dict = scores.to_dict()
-        scores_dict["percent_correct"] = scores.total / scores.possible
+        
+        scores_dict["percent_correct"] = round(scores.total / scores.possible,4)
+        scores_dict["total_points_earned"] = scores.total
 
+        pts_poss_dict = {t: [scores_dict[t]["possible"]] if type(scores_dict[t]) == dict else scores_dict[t] for t in scores_dict}
```
Comment (chrispyles) [inline otter/grade/containers.py:230]: The else clause in this dict comprehension is unnecessary. You can change this to

```
{t: [scores_dict[t]["possible"]] for t in scores_dict}
```

since each value should always be a dict.

And can you also do the same for `scores_dict` on the next line?

File: otter/grade/containers.py [hunk 2]
```diff
@@ -218,12 +223,18 @@ def kill_container():
...
         scores_dict = {t: [scores_dict[t]["score"]] if type(scores_dict[t]) == dict else scores_dict[t] for t in scores_dict}
+        pts_poss_dict["file"] = POINTS_POSSIBLE_LABEL
+        pts_poss_dict["percent_correct"] = "--"
```
Comment (chrispyles) [inline otter/grade/containers.py:233]: This will interfere with libraries like pandas automatically parsing the dtype of this column as numeric. Can you set this to 0 or `NA` (which pandas reads as NaN) instead?

File: otter/grade/containers.py [hunk 3]
```diff
@@ -234,4 +245,7 @@ def kill_container():
         os.remove(results_path)
         os.remove(temp_subm_path)
 
-    return df
+    return {
```
Comment (chrispyles) [inline otter/grade/containers.py:248]: (Optional) Perhaps a better way of doing this, so we don't need to construct the `df_pts_poss` dataframe every time would be to have the future return `scores` and then write a helper method to turn a list of `GradingResults` objects into a dataframe that the caller can use to convert them. This refactor is optional though -- if you don't have the time then this is fine as-is.

File: otter/grade/utils.py [hunk 4]
```diff
@@ -8,7 +8,7 @@
 
 
 OTTER_DOCKER_IMAGE_NAME = "otter-grade"
-
+POINTS_POSSIBLE_LABEL="points-per-question"
```
Comment (chrispyles) [inline otter/grade/utils.py:12]: nit: please add back the second blank line

File: otter/grade/utils.py [hunk 5]
```diff
@@ -8,7 +8,7 @@
...
```
Comment (chrispyles) [inline otter/grade/utils.py:11]: nit: please put spaces around the `=`

File: test/test_grade/test_integration.py [hunk 6]
```diff
@@ -183,6 +187,18 @@ def test_single_notebook_grade(mocked_launch_grade):
         "q2b": 2.0,
         "q7": 1.0,
         "percent_correct": 1.0,
```
Comment (chrispyles) [inline test/test_grade/test_integration.py:189]: Please update this value to reflect the value you set in containers.py

File: test/test_grade/test_integration.py [hunk 7]
```diff
@@ -228,6 +244,18 @@ def test_config_overrides(mocked_launch_grade):
...
```
Comment (chrispyles) [inline test/test_grade/test_integration.py:246]: Please update this value to reflect the value you set in containers.py

File: test/test_grade/test_integration.py [hunk 8]
```diff
@@ -138,6 +139,9 @@ def test_notebooks_with_pdfs(expected_points):
     # read the output and expected output
     df_test = pd.read_csv("test/final_grades.csv")
 
+    #remove points possible row
+    df_test = df_test[df_test['file'] != POINTS_POSSIBLE_LABEL]
```
Comment (chrispyles) [inline test/test_grade/test_integration.py:143]: Please update this test to also test that the points possible row has the correct values.

**Review by chrispyles — CHANGES_REQUESTED** [review]

File: otter/grade/containers.py (hunk 1 above)
Comment (sean-morris) [reply otter/grade/containers.py]: I moved the percent_correct and points_possible attributes to after the dictionary creation to avoid having to check to see if the scores_dict[t] was a dictionary. I added this in on the last PR which makes no sense. I am pulling the scores and points possible first from the scores_dict and then adding in the totals now.

File: test/test_grade/test_integration.py (hunk 8 above)
Comment (sean-morris) [reply test/test_grade/test_integration.py]: Removed these lines to test the points possible row has correct values

File: otter/grade/utils.py [hunk 9]
```diff
@@ -9,6 +9,13 @@
 
 OTTER_DOCKER_IMAGE_NAME = "otter-grade"
 
+POINTS_POSSIBLE_LABEL = "points-per-question"
+
+KEY_SCORES_DICT_FILE = "file"
+
+KEY_SCORES_DICT_TTL_PTS = "total_points_earned"
+
+KEY_SCORES_DICT_PCT_CRT = "percent_correct"
```
Comment (chrispyles) [inline otter/grade/utils.py:18]: Please rename the `KEY_SCORES_DICT_*` vars to something more readable, e.g. `KEY_SCORES_DICT_FILE` -> `SCORES_DICT_FILE_KEY`, `KEY_SCORES_DICT_PCT_CRT` -> `SCORES_DICT_PERCENT_CORRECT_KEY`. Not concerned about length of the var names but the abbreviations are somewhat opaque.

File: otter/grade/containers.py [hunk 10]
```diff
@@ -117,9 +117,28 @@ def launch_containers(
...
+    pts_poss_df = pd.DataFrame(pts_poss_dict)
+    full_df.append(pts_poss_df)
+    for f in finished_futures[0]:
+        scores_dict = f.result().to_dict()
+        scores_dict = {t: [scores_dict[t]["score"]] for t in scores_dict}
+        scores_dict[KEY_SCORES_DICT_PCT_CRT] = round(f.result().total / f.result().possible,4)
+        scores_dict[KEY_SCORES_DICT_TTL_PTS] = f.result().total
+        scores_dict[KEY_SCORES_DICT_FILE] = f.result().file
+        df_scores = pd.DataFrame(scores_dict)
+        full_df.append(df_scores)
```
Comment (chrispyles) [inline otter/grade/containers.py:140]: Can you put this logic for converting a `list[GradingResults]` to a dataframe into a helper function in otter/grade/utils.py and use it here? Having all of this logic here makes the code hard to follow.

```python
def merge_scores_to_df(scores: List[GradingResults]) -> pd.DataFrame:
    """
    Convert a list of ``GradingResults`` objects to a scores dataframe, including a row
    with the total point values for each question.

    Args:
        scores (``list[otter.test_files.GradingResults]``): the score objects to merge

    Returns:
        ``pd.DataFrame``: the scores dataframe
    """
    ...
```

File: otter/grade/utils.py (hunk 4 above)
Comment (chrispyles) [reply otter/grade/utils.py:19]: looks like the second whitespace line is gone again

File: test/test_grade/test_integration.py
Comment (chrispyles) [inline test/test_grade/test_integration.py:1]: To prevent regressions, can you add an explicit assertion to `test_notebooks_with_pdfs` that checks that there is a row in `df_test` whose file name is `POINTS_POSSIBLE_LABEL` with a comment like `# check that the row with point totals for each question exists`?

**Review by chrispyles — CHANGES_REQUESTED** [review]

File: otter/grade/__init__.py [hunk 11]
```diff
@@ -129,11 +129,11 @@ def main(
     # Merge dataframes
     output_df = merge_csv(grade_dfs)
     cols = output_df.columns.tolist()
-    output_df = output_df[cols[-1:] + cols[:-1]]
+    output_df = output_df[[KEY_SCORES_DICT_FILE] + cols[:-3] + [KEY_SCORES_DICT_TTL_PTS, KEY_SCORES_DICT_PCT_CRT]]
```
Comment (chrispyles) [inline otter/grade/__init__.py:132]: rather than hardcoding the index in `cols[:-3]` can we instead do something like

```python
question_cols = sorted(c for c in cols if c not in {KEY_SCORES_DICT_FILE, KEY_SCORES_DICT_TTL_PTS, KEY_SCORES_DICT_PCT_CRT})
output_df = output_df[[KEY_SCORES_DICT_FILE, *question_cols, KEY_SCORES_DICT_TTL_PTS, KEY_SCORES_DICT_PCT_CRT]]
```

Comment (sean-morris) [reply otter/grade/__init__.py:132]: That is solid. Thanks. I didn't like that hard-coded moment.

**Review by chrispyles — APPROVED** [review]
